### PR TITLE
[DistillationTrainer refactor] Remove the local sparse top-1 loss path from the base trainer

### PR DESCRIPTION
**Pull request by qgallouedec**
*Plan step 1.5 — stacked on PR #6454. Part of #6449.*

With the server path extracted (#6454), the base trainer no longer needs the mixed top-1 support path: a local teacher always has full logits, so it computes the exact generalized JSD/KL directly.

- `compute_loss`'s local branch now always calls `generalized_jsd_loss` (which still honours `loss_top_k` for the top-k approximation). The `beta > 0 and loss_top_k == 1` special case and its `completion_tokens` slice are gone.
- `_compute_local_sparse_top_1_divergence_loss` is deleted.
- `_compute_sparse_top_1_divergence_loss` and `_get_reverse_kl_top_1_tokens` become **server-only**, so they move to `ServerDistillationTrainer` (defined directly on it, no longer inherited).
- `reverse_kl_top_1_mode` only drove those helpers, so it moves from `DistillationConfig` to `ServerDistillationConfig` (field + "must be one of" validation); the `self.reverse_kl_top_1_mode` assignment moves from the base `__init__` to the subclass `__init__`. Its stale mentions in the `loss_top_k` docs are trimmed.
- The matching config test moves to the server test file.

Base `compute_loss` local path is now a single `generalized_jsd_loss` call.

### Verification done without pytest

- Base trainer/config no longer expose the sparse helpers or `reverse_kl_top_1_mode`; the subclass defines the two helpers in its own `__dict__`; the server config owns `reverse_kl_top_1_mode` and rejects an invalid value.
- The server sparse loss (`beta=1`, ragged batch) computes to the exact same value as before the relocation (`0.2492382…`), finite forward+backward.

### Risk

Medium. Behaviour-preserving for the base (the removed local sparse path was a top-1 approximation users could opt into via `loss_top_k=1`; it now routes through `generalized_jsd_loss(top_k=1)`, which uses the union support rather than the reverse-mode support — a slightly different top-1 objective). Worth confirming no test pinned the old local top-1 numbers. The server path is behaviour-identical.

### Amendment (post-pytest)

Also removed `TestServerVsLocalTeacher` from the server test file. It cross-checked the *local* sparse top-1 loss against the server's — a comparison whose local half this PR deletes, so the test was passing `loss_top_k=1` to a `DistillationConfig` that no longer accepts it. Server-construction end-to-end stays covered by `TestServerDistillationTrainerRaggedGrad`.

---

> [!NOTE]
> **Medium Risk**
> Local training with loss_top_k=1 and mixed beta no longer uses the old sparse top-1 objective (union top-k via generalized_jsd_loss instead), which can change loss values for existing configs; server behavior is intended to stay the same.
> 
> **Overview**
> Removes the **local sparse top-1** distillation path from `DistillationTrainer` and `DistillationConfig`. The local `compute_loss` branch always calls **`generalized_jsd_loss`** (still honoring `loss_top_k`), instead of the former `beta > 0 && loss_top_k == 1` shortcut that built a teacher-top-1 / reverse-token support set.
> 
> **`reverse_kl_top_1_mode`** and the sparse helpers (`_get_reverse_kl_top_1_tokens`, `_compute_sparse_top_1_divergence_loss`, `_compute_local_sparse_top_1_divergence_loss`) move to **`ServerDistillationTrainer`** / **`ServerDistillationConfig`**, where sparse server logprobs still need them.
> 
> Tests drop the base-config validation for `reverse_kl_top_1_mode`, add it on the server config, remove **`TestServerVsLocalTeacher`** (it depended on the deleted local sparse path), and relax the self-distillation **train_loss** floor to tolerate tiny negative float noise.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a41d5df8c076170cf8b6a37c50772a21e09cd3bf. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_trainer.py [hunk 1]
```diff
@@ -1237,31 +1146,21 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
-                student_logits=student_logits,
-                teacher_logits=teacher_logits,
-                labels=labels,
-                beta=self.beta,
-                temperature=self.temperature,
-                top_k=self.loss_top_k,
-                add_tail=self.loss_add_tail,
-                num_items_in_batch=num_items_in_batch,
-            )
+        loss = self.generalized_jsd_loss(
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:1154]: IWOPDTrainer is a frozen snapshot on purpose; its docstring flags it as not tracking this trainer's later changes. The consistency mandate is for actively-maintained duplicates; a deliberately-frozen fork is exempt by construction (splitting it out was precisely to stop propagating these edits). Its reason for existing is distillation_objective="iw_opd", not the legacy jsd path. Leaving as-is.

**Review by cmpatino — COMMENTED** [review]
The simplification is great, but it introduces a nuance on the top-1 scenario when `beta=1` and `loss_top_k=1`.

The PR removes the option to choose whether the student token should be the one with the highest log-probability (argmax) or the one sampled during decoding (sampled). With the changes in the PR, we only support argmax but the sampled option is the preferred choice because it yields better results. It's also the method used by [Thinking Machines](https://thinkingmachines.ai/blog/on-policy-distillation/) and frameworks like [verl](https://verl.readthedocs.io/en/latest/algo/opd.html#distillation-distillation-loss-loss-mode-str).

What if instead of dropping it entirely we keep:
- `loss_top_k=0`: exact full-vocabulary loss.
- `loss_top_k=1`: sampled completion token for the reverse component.
- Remove `reverse_kl_top_1_mode` entirely.

And handle the logic with something like 

```python
if beta == 0:
    support = teacher_logits.topk(top_k, dim=-1).indices
elif top_k == 1:
    reverse_top = completion_tokens.unsqueeze(-1)
    support = reverse_top if beta == 1 else torch.cat([teacher_top, reverse_top], dim=-1)
```

That way we can preserve the case for the reverse KL with the sampled token behavior with a single `compute_loss` path.

**Comment by cmpatino** [comment]
I take back my earlier comment about trying to keep the top-1 feature. See full comment here https://github.com/huggingface/trl/pull/6456#issuecomment-5023582170
